### ISSUE TITLE
MFA Yubikey not working in Google Chrome and Safari

### ISSUE THREAD

**Issue by stevenGravy** [gravitational/teleport#8017]
### Description

**What happened**:
Unable to use a yubikey in google chrome in  Version 92.0.4515.159 (Official Build) (64-bit)

![image](https://user-images.githubusercontent.com/60704961/130335291-696bb508-5867-4970-8560-0b62b2794ef3.png)

This does work from the command line and firefox browser.

**What you expected to happen**:

Yubikey works as a MFA option in Google chrome

### Reproduction Steps
As minimally and precisely as possible, describe step-by-step how to reproduce the problem.
1.  ON a ubuntu 20.04 install google chrome
2.  On a MFA enabled configuration attempt to use yubikey for a node access in Google chrome

Note on this same machine yubikey does work in AWS access usage.

### Server Details
-  Teleport version (run `teleport version`): 7.0.3
- Server OS (e.g. from `/etc/os-release`): ubuntu 20.04
- Where are you running Teleport? (e.g. AWS, GCP, Dedicated Hardware): AWS
- Additional details:

### Client Details
- Computer OS (e.g. Linux, macOS, Windows):linux ubuntu 20.04
- Browser version (for UI-related issues): Google chrome Version 92.0.4515.159 (Official Build) (64-bit)

**Comment by codingllama** [comment]
It's been a couple days since I've been looking into this one, so let me try to unpack the initial ticket and summarize what I've found out so far.

First, the easy ones:
* tsh is not a factor in this particular bug - as we'll see below, it seems tied to browser appId/origin validation
* Firefox, thankfully, seems to work for all relevant scenarios

Safari doesn't seem to support U2F out of the box ([a fact the code seems acutely aware of](https://github.com/gravitational/webapps/blob/master/packages/teleport/src/services/auth/auth.js#L30)). This seems to have always been the case and is an issue bigger than this particular bug. The best answer for this is to [move to WebAuthn](https://github.com/gravitational/teleport/issues/6478), a work that I've undertaken myself and that should land in the next major. Incidentally, I suspect WebAuthn might be a good answer to the entire bug, but let's try to get a better answer before we yield.

Now, onto the scenario of the bug itself, while talking to @stevenGravy and looking at his cluster, I found the configurations below interesting:

1. authentication.type is set to oidc, which means most users clear login sidestepping U2F issues
2. authentication.u2f is sensibly configured and looks correct to me - no low-hanging fruit on App ID or Facets.
3. a particular node role has require_session_mfa=true, allowing us to hit the bug

If you were to login in the cluster above using Chrome (or any Chrome-based browsers) you'd fail login right away with the following error, which is also characteristic of an App ID mismatch:

<img width="407" alt="Screen Shot 2021-09-29 at 15 11 21" src="https://user-images.githubusercontent.com/12500300/135325308-4901d1df-b00d-4c87-bb50-6373caf1a176.png">

(The above error happens without any user interaction, another interesting data point.)

A comparison of Firefox and Chrome challenge payloads shows that they are receiving essentially identical data (in particular, a correct app ID). Digging further down into Chrome I am led to believe that it is, for some unfathomable reason, failing origin validation - in other words, it is deciding that `https://foo.somedomain.com` is not allowed to use the app ID `https://foo.somedomain.com:443`, even before reaching the U2F device. This seem to happen in the depths of the CryptoTokenExtension, far away from code that we can directly influence. If it indeed is a bug, I also suspect the answer might be "upgrade to WebAuthn", hence my earlier comments.

Something that is still bothering me is that I couldn't repro the bug or nail down the conditions that make it happen - I can interact with it in the wild, as in Steven's cluster, but couldn't yet conjure it at will. I have a few ideas to try yet and I'm hoping this may help shed more light in the issue, but this one may be out of our hands.

**Comment by codingllama** [comment]
OK, I did more digging and got a repro together. As it turns out, what one needs to repro it is to use port 443 for the Web UI and the app ID (eg, "auth_service.authentication.u2f.app_id = https://example.com:443").

The root cause is indeed origin validation. If the app ID trivially matches the origin[1] (say, origin and app ID are both "https://example.com"), then authentication is allowed to proceed. In our scenario origin and app ID don't trivially match because of the ":443" in the app ID, so Chrome falls back to fetching the trustedFacets json[2], fails in that too and ultimately denies us access to U2F.

This boils down to the fact that it is correct for the U2F App ID to contain the port, _unless_ it is a default port (aka 443)[3] [4]. There are decisions we could have made differently, in particular for multi-facet apps, but going down this rabbit hole won't fix the issue as seen in the wild.

"So what now?", you ask. I can see a couple mitigations/solutions for folks suffering from this bug:

1. Upgrade to WebAuthn once it's available.

    WebAuthn should land soon :tm: and I have verified that it fixes this issue. Without going too deep, WebAuthn sets the U2F app ID through a different mechanism and doesn't seem to hit the same validation wall that we do in Chrome. Plus, we get Safari support to boot.

2. Fallback to Firefox

    I wish I had a better answer, but the fact is that Firefox works and we can rely on it until (1) is available.

If you are starting a new cluster and using port 443 for the interface, save yourself the trouble and do not add the port to your app ID.

[1] https://github.com/google/u2f-ref-code/blob/e292d20bf5c2a8b144cdb3d54a4c0fa3e4dae07b/u2f-chrome-extension/appid.js#L174

[2] https://fidoalliance.org/specs/fido-u2f-v1.0-ps-20141009/fido-appid-and-facets-ps-20141009.html#trustedfacets-structure

[3] https://developers.yubico.com/U2F/App_ID.html

[4] https://www.npmjs.com/package/u2f-api#user-content-common-problems


__TL;DR:__

* If you are suffering from this issue upgrade to WebAuthn (as soon as is available)
* If you are setting up a new cluster and using port 443 for the Web UI, then _do not_ add the port to your U2F app ID
* If the above is not possible, fallback to Firefox
